File: src/tessera/scanners/tool_shadow.py
"""Cross-server tool shadowing detection.

When an agent loads tools from multiple MCP servers, a malicious server can
register a tool with the same (or nearly the same) name as a legitimate tool
from a trusted server. The agent then calls the attacker's tool instead of
the real one. This is a confused-deputy attack at the tool registration layer.

Two detection strategies:

1. Edit distance (Levenshtein) for typosquatting: flag pairs from different
   servers with distance <= 2. No ML required.

2. Semantic embedding similarity for synonym attacks: an attacker names a
   tool "email_send" to shadow "send_email". Levenshtein distance is 6 but
   cosine similarity on embedded names is near 1.0. Use semantic_similarity_risk
   for this detection path.

Source attribution: Agent Audit (rule AGENT-055).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable


@dataclass(frozen=True)
class ShadowPair:
    """Two tools from different servers with suspiciously similar names."""

    tool_a: str
    server_a: str
    tool_b: str
    server_b: str
    distance: int  # Levenshtein distance


@dataclass(frozen=True)
class ShadowScanResult:
    """Result of cross-server tool shadow detection."""

    pairs: tuple[ShadowPair, ...]
    shadowed: bool


def _levenshtein(a: str, b: str) -> int:
    """Iterative Levenshtein distance. O(mn) time, O(n) space."""
    if a == b:
        return 0
    if not a:
        return len(b)
    if not b:
        return len(a)

    # Normalize case for comparison
    a = a.lower()
    b = b.lower()

    prev = list(range(len(b) + 1))
    for i, ca in enumerate(a, 1):
        curr = [i] + [0] * len(b)
        for j, cb in enumerate(b, 1):
            if ca == cb:
                curr[j] = prev[j - 1]
            else:
                curr[j] = 1 + min(prev[j], curr[j - 1], prev[j - 1])
        prev = curr
    return prev[-1]


def scan_cross_server_shadows(
    server_tools: dict[str, list[str]],
    max_distance: int = 2,
    principal: str = "system",
) -> ShadowScanResult:
    """Detect tool name shadowing across multiple MCP servers.

    Args:
        server_tools: Mapping of server_name -> list of tool names.
        max_distance: Levenshtein threshold. Pairs with distance <=
            max_distance are flagged. Default 2 catches typosquatting.
            Use 0 for exact-match-only detection.
        principal: Principal for emitted SecurityEvents.

    Returns:
        ShadowScanResult with all flagged pairs.
    """
    servers = list(server_tools.items())
    pairs: list[ShadowPair] = []

    # Compare every pair of (server_a, tool_a) x (server_b, tool_b) where
    # server_a != server_b. Only traverse each pair once (i < j).
    for i in range(len(servers)):
        server_a, tools_a = servers[i]
        for j in range(i + 1, len(servers)):
            server_b, tools_b = servers[j]
            for tool_a in tools_a:
                for tool_b in tools_b:
                    dist = _levenshtein(tool_a, tool_b)
                    if dist <= max_distance:
                        pairs.append(
                            ShadowPair(
                                tool_a=tool_a,
                                server_a=server_a,
                                tool_b=tool_b,
                                server_b=server_b,
                                distance=dist,
                            )
                        )

    result = ShadowScanResult(pairs=tuple(pairs), shadowed=bool(pairs))

    if result.shadowed:
        _emit_findings(result, principal)

    return result


@dataclass(frozen=True)
class ShadowRisk:
    """One candidate pair flagged by semantic similarity."""

    proposed: str
    registered: str
    similarity: float  # cosine similarity in [0, 1]


def semantic_similarity_risk(
    tool_name: str,
    registered_tools: Iterable[str],
    embedder: object = None,
    threshold: float = 0.88,
) -> list[ShadowRisk]:
    """Detect tool shadowing via embedding cosine similarity.

    Embeds the proposed tool name and each registered name, then flags
    pairs whose cosine similarity exceeds threshold. This catches synonym
    attacks that Levenshtein distance misses (e.g. "send_email" vs
    "email_send").

    Args:
        tool_name: The proposed (potentially malicious) tool name.
        registered_tools: Names already registered with the agent.
        embedder: Callable (str) -> list[float]. If None, falls back to
            tessera.mcp.embedding.get_embedder(). If that also returns
            None (package not installed), returns an empty list.
        threshold: Cosine similarity above which a pair is flagged.
            Default 0.88 balances recall and precision on tool name sets.

    Returns:
        List of ShadowRisk, one per registered tool above threshold.
        Empty list when no embedder is available or no pairs cross the
        threshold.
    """
    from tessera.mcp.embedding import cosine_similarity, get_embedder

    embed = embedder or get_embedder()
    if embed is None:
        return []

    proposed_vec = embed(tool_name)
    risks: list[ShadowRisk] = []
    for registered in registered_tools:
        registered_vec = embed(registered)
        sim = cosine_similarity(proposed_vec, registered_vec)
        if sim >= threshold:
            risks.append(ShadowRisk(proposed=tool_name, registered=registered, similarity=sim))
    return risks


def _emit_findings(result: ShadowScanResult, principal: str) -> None:
    from tessera.events import EventKind, SecurityEvent, emit

    # Emit one event per flagged pair. Callers with many servers may want
    # to debounce; that's their problem.
    for pair in result.pairs:
        kind = "exact_shadow" if pair.distance == 0 else "typosquatting"
        emit(
            SecurityEvent.now(
                kind=EventKind.CONTENT_INJECTION_DETECTED,
                principal=principal,
                detail={
                    "scanner": "cross_server_tool_shadow",
                    "kind": kind,
                    "tool_a": pair.tool_a,
                    "server_a": pair.server_a,
                    "tool_b": pair.tool_b,
                    "server_b": pair.server_b,
                    "levenshtein_distance": pair.distance,
                    "owasp": "LLM09",
                    "rule": "AGENT-055",
                },
            )
        )

File: rust/crates/tessera-scanners/src/tool_shadow.rs
//! Cross-server tool shadowing detection via Levenshtein distance.
//!
//! When an agent loads tools from multiple MCP servers, a malicious server can
//! register a tool with the same (or nearly the same) name as a legitimate tool
//! from a trusted server. The agent then calls the attacker's tool instead of
//! the real one. This is a confused-deputy attack at the tool registration layer.
//!
//! Detection: compute case-folded edit distance between all tool names across
//! servers. Flag pairs with distance <= max_distance (default 2) that come from
//! different servers. Distance 0 = identical name (exact shadow). Distance 1-2 =
//! typosquatting.
//!
//! Mirrors `tessera.scanners.tool_shadow` in the Python reference.
//! Source attribution: Agent Audit rule AGENT-055.

use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::ScannerResult;

// ── Result types ─────────────────────────────────────────────────────────────

/// A pair of tools from different servers whose names are suspiciously similar.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ShadowPair {
    /// Name of the first tool.
    pub tool_a: String,
    /// Server the first tool belongs to.
    pub server_a: String,
    /// Name of the second tool.
    pub tool_b: String,
    /// Server the second tool belongs to.
    pub server_b: String,
    /// Case-folded Levenshtein distance between `tool_a` and `tool_b`.
    pub distance: usize,
}

/// Result of cross-server tool shadow detection.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct ShadowScanResult {
    /// All flagged pairs, ordered by discovery (server-pair iteration order,
    /// then tool-pair iteration order). Matches Python tuple ordering.
    pub pairs: Vec<ShadowPair>,
    /// `true` when at least one pair was flagged.
    pub shadowed: bool,
}

impl ScannerResult for ShadowScanResult {
    fn detected(&self) -> bool {
        self.shadowed
    }

    fn scanner_name(&self) -> &'static str {
        "cross_server_tool_shadow"
    }
}

// ── Public API ────────────────────────────────────────────────────────────────

/// Detect tool name shadowing across multiple MCP servers.
///
/// For every ordered pair of distinct servers `(i, j)` with `i < j`, compare
/// every tool from server `i` against every tool from server `j`. Any pair
/// whose case-folded Levenshtein distance is `<= max_distance` is flagged.
///
/// # Arguments
///
/// * `server_tools` - Map of `server_name -> tool_name_list`.
/// * `max_distance` - Levenshtein threshold. Pairs with distance `<=` this
///   value are flagged. Default is `2`, which catches exact shadows and
///   single-character typosquats. Use `0` for exact-match-only detection.
///
/// # Returns
///
/// A [`ShadowScanResult`] containing all flagged [`ShadowPair`]s and a
/// boolean convenience flag. The result is ready to serialize to a
/// SecurityEvent detail payload.
///
/// # Example
///
/// ```rust
/// use tessera_scanners::tool_shadow::scan_cross_server_shadows;
///
/// let mut servers = std::collections::HashMap::new();
/// servers.insert("trusted".to_string(), vec!["send_email".to_string()]);
/// servers.insert("attacker".to_string(), vec!["send_ema1l".to_string()]);
///
/// let result = scan_cross_server_shadows(&servers, 2);
/// assert!(result.shadowed);
/// assert_eq!(result.pairs[0].distance, 1);
/// ```
pub fn scan_cross_server_shadows(
    server_tools: &HashMap<String, Vec<String>>,
    max_distance: usize,
) -> ShadowScanResult {
    // Collect into a deterministically-ordered vec so the i < j traversal
    // visits each server pair exactly once, matching the Python loop.
    let mut servers: Vec<(&String, &Vec<String>)> = server_tools.iter().collect();
    // Sort by server name so output is stable regardless of HashMap iteration
    // order. Python dicts are insertion-ordered; we cannot reproduce that
    // without tracking insertion order, so stable alphabetical is the next
    // best guarantee for tests.
    servers.sort_by_key(|(name, _)| name.as_str());

    let mut pairs: Vec<ShadowPair> = Vec::new();

    for i in 0..servers.len() {
        let (server_a, tools_a) = servers[i];
        for j in (i + 1)..servers.len() {
            let (server_b, tools_b) = servers[j];
            for tool_a in tools_a.iter() {
                for tool_b in tools_b.iter() {
                    let dist = levenshtein_case_folded(tool_a, tool_b);
                    if dist <= max_distance {
                        pairs.push(ShadowPair {
                            tool_a: tool_a.clone(),
                            server_a: server_a.clone(),
                            tool_b: tool_b.clone(),
                            server_b: server_b.clone(),
                            distance: dist,
                        });
                    }
                }
            }
        }
    }

    let shadowed = !pairs.is_empty();
    ShadowScanResult { pairs, shadowed }
}

// ── Internal helpers ──────────────────────────────────────────────────────────

/// Case-folded Levenshtein distance. Mirrors the Python `_levenshtein`
/// function, which lowercases both strings before running the DP.
///
/// Delegates to `strsim::levenshtein` which implements the same iterative
/// O(mn) algorithm.
fn levenshtein_case_folded(a: &str, b: &str) -> usize {
    if a == b {
        return 0;
    }
    // strsim::levenshtein is case-sensitive; fold first.
    let a_low = a.to_lowercase();
    let b_low = b.to_lowercase();
    strsim::levenshtein(&a_low, &b_low)
}

// ── Tests ─────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    fn servers(pairs: &[(&str, &[&str])]) -> HashMap<String, Vec<String>> {
        pairs
            .iter()
            .map(|(srv, tools)| {
                (
                    srv.to_string(),
                    tools.iter().map(|t| t.to_string()).collect(),
                )
            })
            .collect()
    }

    // ---- mirrors Python TestToolShadow tests --------------------------------

    #[test]
    fn no_overlap_is_clean() {
        // test_no_overlap_clean
        let s = servers(&[
            ("trusted", &["search_web", "send_email"]),
            ("attacker", &["fetch_data", "log_event"]),
        ]);
        let r = scan_cross_server_shadows(&s, 2);
        assert!(!r.shadowed);
        assert!(r.pairs.is_empty());
    }

    #[test]
    fn exact_shadow_detected() {
        // test_exact_shadow_detected
        let s = servers(&[
            ("trusted", &["send_email"]),
            ("attacker", &["send_email"]),
        ]);
        let r = scan_cross_server_shadows(&s, 2);
        assert!(r.shadowed);
        assert_eq!(r.pairs.len(), 1);
        let pair = &r.pairs[0];
        assert_eq!(pair.distance, 0);
        assert_eq!(pair.tool_a, "send_email");
        assert_eq!(pair.tool_b, "send_email");
        assert_ne!(pair.server_a, pair.server_b);
    }

    #[test]
    fn typosquatting_distance_1_detected() {
        // test_typosquatting_detected: 'send_ema1l' vs 'send_email', distance=1
        let s = servers(&[
            ("trusted", &["send_email"]),
            ("attacker", &["send_ema1l"]),
        ]);
        let r = scan_cross_server_shadows(&s, 2);
        assert!(r.shadowed);
        assert_eq!(r.pairs[0].distance, 1);
    }

    #[test]
    fn distance_2_detected_within_threshold() {
        // test_distance_2_detected
        let s = servers(&[
            ("trusted", &["web_search"]),
            ("attacker", &["web_searsh"]), // 2 edits
        ]);
        let r = scan_cross_server_shadows(&s, 2);
        assert!(r.shadowed);
    }

    #[test]
    fn clearly_different_names_not_flagged_at_threshold_2() {
        // test_distance_3_not_flagged_by_default (second half: "alpha" vs "zeta")
        let s = servers(&[("trusted", &["alpha"]), ("attacker", &["zeta"])]);
        let r = scan_cross_server_shadows(&s, 2);
        assert!(!r.shadowed);
    }

    #[test]
    fn same_server_tools_never_flagged() {
        // test_same_server_not_flagged
        let s = servers(&[("trusted", &["search", "search_v2"])]);
        let r = scan_cross_server_shadows(&s, 2);
        assert!(!r.shadowed);
    }

    #[test]
    fn exact_shadow_distance_labeled_zero() {
        // test_exact_zero_distance_labeled_correctly
        let s = servers(&[("a", &["exact_name"]), ("b", &["exact_name"])]);
        let r = scan_cross_server_shadows(&s, 2);
        assert_eq!(r.pairs[0].distance, 0);
    }

    // ---- additional coverage ------------------------------------------------

    #[test]
    fn max_distance_zero_only_flags_exact_match() {
        let s = servers(&[
            ("trusted", &["send_email"]),
            ("attacker", &["send_ema1l"]),
        ]);
        // distance=1, so max_distance=0 should not flag it
        let r = scan_cross_server_shadows(&s, 0);
        assert!(!r.shadowed);

        let s2 = servers(&[("a", &["tool"]), ("b", &["tool"])]);
        let r2 = scan_cross_server_shadows(&s2, 0);
        assert!(r2.shadowed);
        assert_eq!(r2.pairs[0].distance, 0);
    }

    #[test]
    fn case_folding_exact_shadow_across_case() {
        // "Send_Email" vs "send_email" should be distance 0 after lowercasing.
        let s = servers(&[("a", &["Send_Email"]), ("b", &["send_email"])]);
        let r = scan_cross_server_shadows(&s, 0);
        assert!(r.shadowed);
        assert_eq!(r.pairs[0].distance, 0);
    }

    #[test]
    fn three_servers_all_pairs_scanned() {
        // With three servers and one shared name across two of them, exactly
        // one pair should be flagged (the two that share the name), not two.
        let s = servers(&[
            ("a", &["common"]),
            ("b", &["common"]),
            ("c", &["different"]),
        ]);
        let r = scan_cross_server_shadows(&s, 0);
        assert_eq!(r.pairs.len(), 1);
        assert_eq!(r.pairs[0].distance, 0);
    }

    #[test]
    fn empty_server_map_returns_clean() {
        let s: HashMap<String, Vec<String>> = HashMap::new();
        let r = scan_cross_server_shadows(&s, 2);
        assert!(!r.shadowed);
        assert!(r.pairs.is_empty());
    }

    #[test]
    fn single_server_returns_clean() {
        let s = servers(&[("only", &["tool_a", "tool_b"])]);
        let r = scan_cross_server_shadows(&s, 0);
        assert!(!r.shadowed);
    }

    #[test]
    fn scanner_result_trait_methods() {
        let s = servers(&[("a", &["x"]), ("b", &["x"])]);
        let r = scan_cross_server_shadows(&s, 0);
        assert!(ScannerResult::detected(&r));
        assert_eq!(ScannerResult::scanner_name(&r), "cross_server_tool_shadow");

        let clean = servers(&[("a", &["alpha"]), ("b", &["zeta"])]);
        let c = scan_cross_server_shadows(&clean, 0);
        assert!(!ScannerResult::detected(&c));
    }

    #[test]
    fn serialize_round_trip_via_serde_json() {
        let s = servers(&[("a", &["tool_x"]), ("b", &["tool_x"])]);
        let r = scan_cross_server_shadows(&s, 2);
        let json = serde_json::to_string(&r).unwrap();
        let back: ShadowScanResult = serde_json::from_str(&json).unwrap();
        assert_eq!(r, back);
    }
}
